src/Components/Aspire.Azure.Npgsql.EntityFrameworkCore.PostgreSQL/AspireAzureEFPostgreSqlExtensions.cs

```diff
@@ -0,0 +1,171 @@
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        AzureNpgsqlEntityFrameworkCorePostgreSQLSettings? azureSettings = null;
+
+        builder.EnrichNpgsqlDbContext<TContext>(settings => azureSettings = ConfigureSettings(configureSettings, settings));
+
+        // Enrich should always call ConfigureSettings
+        Debug.Assert(azureSettings != null);
+
+#if NET9_0_OR_GREATER
+        builder.Services.ConfigureDbContext<TContext>(dbContextOptionsBuilder => ConfigureDbContextOptionsBuilder(azureSettings, dbContextOptionsBuilder));
+#else
+        builder.PatchServiceDescriptor<TContext>(dbContextOptionsBuilder =>
+        {
+            ConfigureDbContextOptionsBuilder(azureSettings, dbContextOptionsBuilder);
+        });
+#endif
+    }
+
+    private static AzureNpgsqlEntityFrameworkCorePostgreSQLSettings ConfigureSettings(Action<AzureNpgsqlEntityFrameworkCorePostgreSQLSettings>? userConfigureSettings, NpgsqlEntityFrameworkCorePostgreSQLSettings settings)
+    {
+        var azureSettings = new AzureNpgsqlEntityFrameworkCorePostgreSQLSettings();
+
+        // Copy the values updated by Npgsql integration.
+        CopySettings(settings, azureSettings);
+
+        // Invoke the Aspire configuration.
+        userConfigureSettings?.Invoke(azureSettings);
+
+        // Copy to the Npgsql integration settings as it needs to get any values set in userConfigureSettings.
+        CopySettings(azureSettings, settings);
+
+        return azureSettings;
+    }
+
+    private static void CopySettings(NpgsqlEntityFrameworkCorePostgreSQLSettings source, NpgsqlEntityFrameworkCorePostgreSQLSettings destination)
+    {
+        destination.ConnectionString = source.ConnectionString;
+        destination.DisableHealthChecks = source.DisableHealthChecks;
+        destination.DisableMetrics = source.DisableMetrics;
+        destination.DisableTracing = source.DisableTracing;
+        destination.DisableRetry = source.DisableRetry;
+        destination.CommandTimeout = source.CommandTimeout;
+    }
+
+    private static void ConfigureDbContextOptionsBuilder(AzureNpgsqlEntityFrameworkCorePostgreSQLSettings settings, DbContextOptionsBuilder dbContextOptionsBuilder)
+    {
+        // The connection string requires the username to be provided. Since it will depend on the Managed Identity that is used
+        // we attempt to get the username from the access token.
+
+        var credential = settings.Credential ?? new DefaultAzureCredential();
+
+#pragma warning disable EF1001 // Internal EF Core API usage.
+
+        // Get the connection string from the Npgsql options extension in case it was set using UseNpgsql(connStr) and Enrich()
+        var extensionsConnectionString = dbContextOptionsBuilder.Options.GetExtension<NpgsqlOptionsExtension>()?.ConnectionString;
+
+#pragma warning restore EF1001 // Internal EF Core API usage.
+
+        var dataSourceBuilder = new NpgsqlDataSourceBuilder(settings.ConnectionString ?? extensionsConnectionString);
```

Comment:
If we are on NET9_0_OR_GREATER, we don't need to do it this way. Instead, we can use `npgsqlOptions.ConfigureDataSource` to configure the datasource builder. That route is more composable because it can be configured by multiple pieces of code in the pipeline.

I know it will mean that we will need to split the code more, but I think when we are on 9.0+ we should be using this new preferred method, and when net8 goes EOL, we can drop the old route of creating the `NpgsqlDataSourceBuilder` and setting it on the options.

@roji - Do you have thoughts/opinions here?